File: 901-1000/981-990/981-timeBasedKeyValueStore/timeBasedKeyValueStore.py
import bisect

class TimeMap(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """

        self.hash_set = {}


    def set(self, key, value, timestamp):
        """
        :type key: str
        :type value: str
        :type timestamp: int
        :rtype: None
        """

        if key not in self.hash_set:
            self.hash_set[key] = []

        self.hash_set[key].append((timestamp, value))


    def get(self, key, timestamp):
        """
        :type key: str
        :type timestamp: int
        :rtype: str
        """

        if key in self.hash_set:
            index = bisect.bisect(self.hash_set[key], (timestamp+1, ""))
            return self.hash_set[key][index-1][1] if index else ""

        return ""


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: 901-1000/981-990/981-timeBasedKeyValueStore/timeBasedKeyValueStore.cpp
#include <unordered_map>
#include <map>
#include <string>

using namespace std;

class TimeMap {
public:
    /** Initialize your data structure here. */
    unordered_map<string, map<int, string>> timeMap;

    TimeMap() {
    }

    void set(string key, string value, int timestamp) {
        timeMap[key][timestamp] = value;
    }

    string get(string key, int timestamp) {
        if (timeMap[key].size() > 0) {
            auto iter = timeMap[key].upper_bound(timestamp);
            if (iter == timeMap[key].end())
                return timeMap[key].rbegin()->second;
            else if (iter == timeMap[key].begin())
                return "";
            else
                return (--iter)->second;
        }

        return "";
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */